src/dtree.ml

```diff
@@ -0,0 +1,398 @@
+      to a heuristic.  *)
+  let cmp : line -> line -> int = fun _ _ -> 0
+
+  (** [pick_best m] returns the index of the best column of matrix [m]
+      according to a heuristic. *)
+  let pick_best : t -> int = fun _ -> 0
+
+  (** [is_pattern e t] returns whether a term [t] is considered as a pattern
+      wrt the environment [e] attached to the rule containing [t]. *)
+  let rec is_pattern : term_env array -> term -> bool = fun ar te ->
+    match te with
+    | Patt(None, _, [| |])                          -> false
+    (* ^ Wildcard *)
+    | Patt(Some(i), _, [| |]) when ar.(i) = TE_None -> false
+    (* ^ Linear pattern used in rhs *)
+    | Patt(_, _, _)                                 -> false
+    | Appl(u, _)                                    -> is_pattern ar u
+    (* ^ Should be useless *)
+    | _                                             -> true
+
+  (** [exhausted r] returns whether rule [r] can be further pattern matched or
+      if it is ready to yield the action.  A rule is exhausted when its left
+      hand side is composed only of wildcards. *)
+  let exhausted : rule -> bool = fun { lhs = lh ; env = en ; _ }->
+    let rec loop = function
+      | []                          -> true
+      | x :: _ when is_pattern en x -> false
+      | _ :: xs                     -> loop xs in
+    loop lh
+
+  (** [can_switch_on p k] returns whether a switch can be carried out on
+      column [k] in matrix [p] *)
+  let can_switch_on : t -> int -> bool = fun { values = valu ; _ } k ->
+    let rec loop : rule list -> bool = function
+      | []      -> true
+      | r :: rs ->
+        begin
+          match List.nth_opt r.lhs k with
+          | None    -> loop rs
+          | Some(t) -> is_pattern r.env t
+        end in
+    loop valu
+
+  (** [discard_patt_free m] returns the list of indexes of columns containing
+      terms that can be matched against (discard pattern-free columns). *)
+  let discard_patt_free : t -> int array = fun m ->
+    let ncols = List.fold_left (fun acc { lhs = l ; _ } ->
+        let le = List.length l in
+      if le > acc then le else acc) 0 m.values in
+    let switchable = List.init ncols (fun k ->
+        can_switch_on m k) in
+    let indexes = List.mapi (fun k cm ->
+        if cm then Some(k) else None) switchable in
+    let remaining = List.filter (function
+        | None    -> false
+        | Some(_) -> true) indexes in
+    let unpacked = List.map (function
+        | Some(k) -> k
+        | None    -> assert false) remaining in
+    assert ((List.length unpacked) > 0) ;
+    Array.of_list unpacked
+end
+
+module Pm = Pattmat
+
+(** [appl_leftmost t] returns the leftmost non {!cons:Appl} term of [t]. *)
+let rec appl_leftmost : term -> term = function
+  | Appl(u, _) -> appl_leftmost u
+  | u          -> u
+
+(** [appl_left_depth t] returns the number of successive {!cons:Appl} on the
+    left of the expression. *)
+let rec appl_left_depth : term -> int = function
```

Comment:
What is it for?

Reply:
When matching on a term, always call unfold first (metavariables can be instantiated).

Reply:
Used Basics.get_args in 4f4ad87.